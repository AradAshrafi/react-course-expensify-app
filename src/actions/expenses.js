// import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE

// {
//   description = '',
//   note = '',
//   amount = 0,
//   createdAt = 0
// } = {}

export const addExpense = expense => ({
    type: 'ADD_EXPENSE',
    expense
    // id: uuid(),
});

export const startAddExpense = (expenseData = {}) => {
    return x => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        database
            .ref('expenses')
            .push(expense)
            .then(ref => {
                x(
                    addExpense({
                        id: ref.key,
                        ...expense
                    })
                );
            });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES

export const setExpense = expenses => {
    return {
        type: 'SET_EXPENSES',
        expenses
    };
};

export const startSetExpenses = () => {
    return toDispatch => {
        return database
            .ref('expenses')
            .once('value')
            .then(snapshot => {
                const expenses = [];
                snapshot.forEach(element => {
                    expenses.push({
                        id:element.key,
                        ...element.val()
                    });
                });
                toDispatch(setExpense(expenses));
            });
    };
};
