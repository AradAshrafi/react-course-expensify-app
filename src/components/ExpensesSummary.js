import React from "react";
import { connect } from "react-redux";
import numeral from 'numeral';
import expenseTotal from "../selectors/expenses-total";
import getVisibleExpense from "../selectors/expenses";
import expensesTotal from "../selectors/expenses-total";

export const ExpensesSummary = props => {
    const expenseWord=props.expenses.length===1 ? 'expense' : 'expenses';
    return(
        <h2>
            viewing {props.expenses.length} {expenseWord} totalling{" "}
            {numeral(expensesTotal(props.expenses)/100).format('&0,0.00')}
        </h2>
    )
};

const mapState2Props = state => {
    return {
        expenses: getVisibleExpense(state.expenses , state.filters)
    };
};

export default connect(mapState2Props)(ExpensesSummary);
