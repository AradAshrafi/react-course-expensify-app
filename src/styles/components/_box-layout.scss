.box-layout{
    align-items: center;
    display: flex;
    justify-content:center;
    background-image: url('/images/bg.jpg');
    background-size: cover;
    height: 100vh;
    width:100vw;
}

.box-layout__box{
    background: fade-out(white, 0.15);
    border-radius: 3px;
    padding : $l-size $m-size;
    text-align: center;
    width: 25rem;
}

.box-layout__title{
    margin:0 0 $m-size;
    line-height: 1;
}