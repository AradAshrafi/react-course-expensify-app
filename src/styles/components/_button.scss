.button{
    background-color: $blue;
    color:white;
    border:none;
    display:inline-block;//added for times that we want to see links as buttons
    padding:$s-size;
    line-height: 1;
    // width: 20rem; :-?y did i add this for login ? consider this if u ran into bugs
    font-weight:300;
    font-size: $font-size-large;
    text-decoration: none;
}

.button__link{
    background:none;
}
.button--remove{
    @extend .button ;
    background-color :#888888;
}

//or can use both button and button--remove classNames together -> className = "button button--remove"