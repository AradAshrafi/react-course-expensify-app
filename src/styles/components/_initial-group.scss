.initial-group {
    display: flex;
    // justify-content : center;
    flex-direction: column;
    // align-items: center;
    margin-bottom: $m-size;
    @media (min-width: $desktop-breakpoint) {
        flex-direction: row;
        margin-bottom: $l-size;
    }
}

.initial-group__item {
    // display: inline-block;
    margin-right: $s-size;
    @media (max-width:$desktop-breakpoint) {
        margin-bottom: $s-size;
    }
}
