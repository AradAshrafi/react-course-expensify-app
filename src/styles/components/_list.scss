.list-header{
    background-color: $off-white;
    border:1px solid darken($off-white,10%);
    color:gray;
    padding:$s-size $m-size;
    display:flex;
    justify-content: space-between;
}
.list-item{
    border:1px solid darken($off-white,10%);
    color:$dark-grey;
    text-decoration: none;
    display: flex;
    flex-direction: column;
    border-top:none;
    padding:$s-size;
    transition: background .3s ease;
    &:hover{ //is equal to .list-items:hover
        background-color : $off-white
    }
    @media (min-width : $desktop-breakpoint) {
        flex-direction: row;
        justify-content: space-between;
    }
}
.list-header__mobile{
    @media (min-width : $desktop-breakpoint ) {
        display: none;
    }
}

.list-header__desktop{
    @media (max-width : $desktop-breakpoint + 0.01rem){
        display : none;
    }
}

.list-item__title{
    margin:0;
    word-break:break-all;
}

.list-item__subtitle{
    color:$grey;
    font-size: $font-size-small;
}

.list-item__data{
    margin : $s-size 0 0 0;
    @media (min-width : $desktop-breakpoint) {
        margin:0;
        padding-left:$s-size;
    }
}
.list-item--message{
    align-items:center;
    justify-content: center;
    color:$grey;
}

.list-body{
    margin-bottom:$m-size;
}