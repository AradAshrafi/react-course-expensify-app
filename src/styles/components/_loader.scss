.loader{
    height:100vh;
    width:100vw;
    display:flex;
    align-items: center;
    justify-content: center;
}

.loader__img{
    height:8rem;
    width:8rem;
}