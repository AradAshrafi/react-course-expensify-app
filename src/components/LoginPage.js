import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

// class LoginPage extends React.Component {
// constructor(props){
//     super(props);
// }
// }

const LoginPage = props => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expensify</h1>
            <p>it's time to get your expenses under control.</p>
                <button onClick={props.dispatch(startLogin)} className="button">
                    Login with Google
                </button>
        </div>
    </div>
);

export default connect()(LoginPage);
