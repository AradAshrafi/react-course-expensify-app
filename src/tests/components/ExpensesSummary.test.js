import React from 'react';
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import expenses from '../fixtures/expenses'

test("1 expense", () => {
    const check={expenses:[expenses[0]]}
    const wrapper = shallow(<ExpensesSummary {...check}/>);
    expect(wrapper).toMatchSnapshot()
});
//expenses={check} ghalateee

test(" multiple expenses ", () => {
    const check = {expenses : expenses }
    const wrapper = shallow(<ExpensesSummary {...check} />);
    expect(wrapper).toMatchSnapshot()
});