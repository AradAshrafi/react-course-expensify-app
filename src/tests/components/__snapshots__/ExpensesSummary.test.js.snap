// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` multiple expenses  1`] = `
<h2>
  viewing 
  3
   
  expenses
   totalling
   
  1,141.95
</h2>
`;

exports[`1 expense 1`] = `
<h2>
  viewing 
  1
   
  expense
   totalling
   
  1.95
</h2>
`;
